File: lib/hubclient/transports/IHubTransport.go
package transports

import (
	"crypto/tls"
	"crypto/x509"
	"errors"
	"github.com/hiveot/hub/lib/keys"
	"github.com/hiveot/hub/runtime/api"
)

// inbox implementation depends on the underlying transport.
// this is the commonly used name for it.
const MessageTypeINBOX = "_INBOX"

// ISubscription interface to underlying subscription mechanism
type ISubscription interface {
	Unsubscribe() error
}

type ConnectionStatus string

const (
	// Connecting attempting a (re)connection
	Connecting ConnectionStatus = "connecting"
	// Connected and authenticated successful
	Connected ConnectionStatus = "connected"
	// ConnectFailed after failure to connect
	// Only used if retry has given up
	ConnectFailed ConnectionStatus = "connectFailed"
	// Disconnected by client or not yet connected
	Disconnected ConnectionStatus = "disconnected"
	// Expired authentication token
	Expired ConnectionStatus = "expired"
	// Unauthorized login name or password
	Unauthorized ConnectionStatus = "unauthorized"
)

// MessageTypeINBOX special inbox prefix for RPCs
// reserved event and action names
const (
// MessageTypeAction = "action"
// MessageTypeConfig = "config"
// MessageTypeEvent  = "event"
// MessageTypeRPC    = "rpc"
// MessageTypeINBOX  = "_INBOX"
// EventTypeTD       = "$td"
// EventTypeProps    = "$properties"
)

var ErrorUnauthorized = errors.New(string(Unauthorized))

type HubTransportStatus struct {
	// URL of the hub
	HubURL string
	// CA used to connect
	CaCert *x509.Certificate
	// The transport core, eg mqtt or nats
	Core string
	// the client ID to identify as
	ClientID string

	// The current connection status
	ConnectionStatus ConnectionStatus
	// The last connection error message, if any
	LastError error

	// flags indicating the supported protocols
	SupportsCertAuth     bool
	SupportsPasswordAuth bool
	SupportsKeysAuth     bool
	SupportsTokenAuth    bool
}

// IHubTransport defines the interface of the transport client that connects to a messaging server.
type IHubTransport interface {

	// ConnectWithCert connects to the server using a client certificate.
	// This authentication method is optional
	ConnectWithCert(kp keys.IHiveKey, cert *tls.Certificate) (token string, err error)

	// ConnectWithPassword connects to the messaging server using password authentication.
	//  loginID is the client's ID (typically consumers)
	//  password is created when registering the user with the auth service.
	//
	// This authentication method must be supported by all transports
	ConnectWithPassword(password string) error

	// ConnectWithJWT connects to the messaging server using an authentication token
	// and pub/private keys provided when creating an instance of the hub client.
	//  token is created by the auth service.
	//
	// This authentication method must be supported by all transports
	ConnectWithJWT(token string) error

	// CreateKeyPair returns a new set of serialized public/private key pair.
	//  serializedKP contains the serialized public/private key pair
	//  pubKey contains the serialized public key to be shared
	CreateKeyPair() (kp keys.IHiveKey)

	// Disconnect from the messaging server.
	// This removes all subscriptions.
	Disconnect()

	// GetStatus returns the current transport connection status
	GetStatus() HubTransportStatus

	// PubAction publishes an action request and waits for a response.
	//	thingID for whom the action is intended
	//	key ID or method name of the action
	//  payload with serialized message to publish
	//  returns a delivery status with serialized response message if delivered
	PubAction(thingID string, key string, payload []byte) api.DeliveryStatus

	// PubEvent publishes an event style message without waiting for a response.
	//	thingID whose event is published
	//	key ID of the event
	//	payload with serialized message to publish
	PubEvent(thingID string, key string, payload []byte) api.DeliveryStatus

	// SetConnectHandler sets the notification handler of connection status changes
	SetConnectHandler(cb func(status HubTransportStatus))

	// SetMessageHandler set the single handler that receives all subscribed messages
	// and messages directed at this client.
	//
	// Note that Agents receive actions with a thingID that does not have the agent
	// prefix as the agent prefix is consumer facing. Things from agents are physical
	// things while digitwin Things are virtual Things with a different ID.
	//
	// Consumers can only interact with the digital twin.
	//
	// See also 'Subscribe' to set the things this client receives messages for.
	SetMessageHandler(cb api.MessageHandler)

	// Subscribe adds a subscription for one or more events from the thingID.
	//
	// This is for events only. Actions directed to this client are automatically passed
	// to this client's messageHandler. The TD documents published by this agent have
	// their ThingID associated with the agent using this transport.
	//
	//
	// Events will be passed to the event handler.
	// This is pretty coarse grained.
	// Subscriptions remain in effect when the connection with the messaging server is interrupted.
	//  thingID is the ID of the Thing whose events to receive or "" for events from all things
	Subscribe(thingID string) error

	// Unsubscribe removes a previous event subscription.
	// No more events or requests will be received after Unsubscribe.
	Unsubscribe(thingID string)
}

File: libjs/src/hubclient/transports/IHubTransport.ts
// ISubscription interface to underlying subscription mechanism
import type { IHiveKey } from "@hivelib/keys/IHiveKey";


export enum ConnectionStatus {
    Connected = "connected",
    Connecting = "connecting",
    Disconnected = "disconnected",
}

export enum ConnInfo {
    // connection successful
    Success = "success",
    // ConnInfoUnauthorized credentials invalid
    Unauthorized = "unauthorized",
    // ConnInfoUnreachable unable to reach the server during the initial connection attempt
    Unreachable = "unreachable",
    // ConnInfoServerDisconnected a server disconnect message was received.
    ServerDisconnected = "serverDisconnected",
    // NetworkDisconnected connection has dropped. Caused by disconnecting the network somewhere
    NetworkDisconnected = "networkDisconnected",
    // NotConnected means the client has yet to connect
    NotConnected = "notConnected"
}

// IHubTransport defines the interface of the message bus transport used by
// the hub client.
export interface IHubTransport {
    // addressTokens returns the address separator and wildcard tokens used by the transport
    // @result sep is the address separator. eg "." for nats, "/" for mqtt and redis
    // @result wc is the address wildcard. "*" for nats, "+" for mqtt
    // @result rem is the address remainder. "" for nats; "#" for mqtt
    addressTokens(): { sep: string, wc: string, rem: string };

    // ConnectWithPassword connects to the messaging server using password authentication.
    // @param loginID is the client's ID
    // @param password is created when registering the user with the auth service.
    connectWithPassword(password: string): Promise<void>;

    // ConnectWithToken connects to the messaging server using an authentication token
    // and pub/private keys provided when creating an instance of the hub client.
    // @param key is the key generated with createKey.
    // @param token is created by the auth service.
    connectWithToken(key: IHiveKey, token: string): Promise<void>;

    // CreateKeyPair returns a new key for authentication and signing.
    // @returns key contains the public/private key pair.
    createKeyPair(): IHiveKey;

    // Disconnect from the message bus.
    disconnect(): void;

    // PubEvent publishes event type messages and returns immediately. 
    // @param address to publish on
    // @param payload with serialized message to publish
    pubEvent(address: string, payload: string): Promise<void>;

    // PubRequest publishes an RPC request and waits for a response.
    // @param address to publish on
    // @param payload with serialized message to publish
    // @returns reply with serialized response message
    pubRequest(address: string, payload: string): Promise<string | boolean>;

    // set handler that is notified of changes in connection status and an error in 
    // case of an  unintentional disconnect.
    // 
    // This handler is intended for updating presentation of the connection status.
    // Do not call connectXyz() in this handler, as a reconnect attempt will be made 
    // after a short delay. If a connection is re-established then the onConnect
    // handler will be invoked.
    //
    //  connected is true if a connection is established or false if disconnected.
    //  info contains human presentable information when available.
    //
    // If a reconnect is to take place with a different password or token then 
    // call disconnect(), followed by connectWithXyz().
    setConnectHandler(handler: (status: ConnectionStatus, info: string) => void): void


    // Set the handler for incoming event-type messages.
    // Event type messages are those that do not contain a reply-to address and correlation data.
    //
    setEventHandler(handler: (addr: string, payload: string) => void): void

    // Set the handler for incoming request-response message.
    // The handler will be invoked if a message is received that contains a reply-to
    // address and a correlation ID. The result of the handler will be send back to
    // the sender. If an exception is thrown then an error will be returned to the sender.
    //
    // Support for request-response messages requires MQTT v5.
    //
    // The result of the handler will be sent as a reply.
    // This requires MQTT v5.
    setRequestHandler(handler: (addr: string, payload: string) => string): void

    // Subscribe adds a subscription for an event or request address.
    // Incoming messages are passed to the event handler or the request handler, depending on whether they
    // have a reply-to address. The event/request handler will handle the routing as this is application specific.
    // Subscriptions remain in effect when the connection with the messaging server is interrupted.
    //
    // The address MUST be constructed using the tokens provided by AddressTokens()
    subscribe(address: string): Promise<void>;

    // unsubscribe removes the address from the subscription list
    unsubscribe(address: string): void;
}
